add lazy store

huiscool opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/daotl/go-datastore/1)
<!-- Reviewable:end -->

Commit:
implements lazy-datastore

Commit:
add a simple lazyDatastore test

NexZhu commented on lazy/lazy.go:
```diff
@@ -0,0 +1,206 @@
+type DSModifier func(d *datastore.Datastore) error
+
+// LazyDatastore wraps a datastore with three states: active, inactive and closed.
+// LazyDatasotre is inited by four DSModifier: init, activate, deactive and closeFunc.
```
deactive -> deactivate

NexZhu replied:
Unify naming:
// LazyDatasotre is inited by four DSModifier: initFn, activateFn, deactiveFn, and closeFn.

NexZhu commented on lazy/lazy.go:
```diff
@@ -0,0 +1,206 @@
+//                       -----------(deactivateFunc)----------
+//                       |                                   |
+//                       V                                   |
+// nil -(initFunc)-> deactive -----(activateFunc)---------> active
```
dactive -> inactive

NexZhu commented on lazy/lazy.go:
```diff
@@ -0,0 +1,206 @@
+	ErrClosed = errors.New("datastore closed")
+)
+
+type DSModifier func(d *datastore.Datastore) error
```
naming: modifier usually refers to static thing, DSAction is better

NexZhu commented on lazy/lazy.go:
```diff
@@ -0,0 +1,206 @@
+}
+
+// LockActive check if l is opened and lock the datastore to active state.
+// If l is opened, prevent datastore's state being changed;
```
unify terms, opend -> activated, open -> activate

NexZhu commented on lazy/lazy.go:
```diff
@@ -0,0 +1,206 @@
+// UnlockActive releases datastore.
+func (l *LazyDatastore) UnlockActive() {
+	if !l.writing {
+		l.rw.RUnlock()
```
There seems to be falty sequence:
Three goroutines: A, B, C
A L74 -> A L81 -> B L74  -> C L101 -> A L82 -> (A L94 & B L75) becomes race condition

Commit:
1. add EnsureActive;
2. unexposed lockActive and unlockActive;
3. rename DSModifier to Action;
4. update some comments;
5. return ErrClosed in a closed LazyDatastore.

NexZhu commented:
Merged
